txstratum/middleware.py

```diff
@@ -0,0 +1,107 @@
+# Copyright (c) Hathor Labs and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import re
+from typing import Awaitable, Callable, Optional
+
+from aiohttp import web
+from aiohttp.typedefs import Handler
+
+from txstratum.utils import is_version_gte
+
+# Middleware factory return type
+Middleware = Callable[[web.Request, Handler], Awaitable[web.StreamResponse]]
+
+# Error message
+VERSION_CHECK_ERROR_MESSAGE = "wallet-version-too-old"
+
+
+def create_middleware_version_check(
+    min_wallet_desktop_version: Optional[str],
+    min_wallet_mobile_version: Optional[str],
+    min_wallet_headless_version: Optional[str],
+) -> Middleware:
+    """Middleware factory."""
+
+    @web.middleware
+    async def version_check(
+        request: web.Request, handler: Handler
+    ) -> web.StreamResponse:
+        """Check wallet versions from user agent."""
+        user_agent = request.headers.get("User-Agent")
+
+        if not user_agent:
+            response = await handler(request)
+            return response
+
+        if min_wallet_desktop_version:
+            # Search user agent for wallet desktop string and get version
+            # Then check if version is the minimum allowed
+            # Example of a wallet desktop user agent
+            # Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) HathorWallet/0.22.1
+            # Chrome/73.0.3683.121 Electron/5.0.13 Safari/537.36 HathorWallet/0.22.1
+            wallet_desktop_regex = r"HathorWallet/(\d+\.\d+.\d+)"
+            if is_version_invalid(
+                wallet_desktop_regex, user_agent, min_wallet_desktop_version
+            ):
+                return web.json_response(
+                    {"error": VERSION_CHECK_ERROR_MESSAGE}, status=400
```

Comment:
Should we add the detected version and the minimum expected version in the response?

Maybe it is useful in some cases, especially in the wallet-headless, that could be logging the received response.

Reply:
I was trying to keep the error string as an "error code", similar to what we have in other responses in the tx mining service but I agree it's useful for logging. 

I've added in another dict field the versions, what do you think? https://github.com/HathorNetwork/tx-mining-service/pull/81/commits/60dd9040f8286ac11ff30a72b038a08678d46bed

Reply:
Great